src/wallet/wallet.cpp

```diff
@@ -4316,127 +4314,51 @@ std::optional<std::pair<
-    assert(maybe_pa != std::nullopt);
-    auto pa = maybe_pa.value();
-
-    return std::make_pair(notePt, pa);
-}
-
-std::optional<std::pair<
-    SaplingNotePlaintext,
-    SaplingPaymentAddress>> CWalletTx::DecryptSaplingNoteWithoutLeadByteCheck(SaplingOutPoint op) const
-{
-    // Check whether we can decrypt this SaplingOutPoint
-    if (this->mapSaplingNoteData.count(op) == 0) {
-        return std::nullopt;
-    }
-
-    auto output = this->vShieldedOutput[op.n];
-    auto nd = this->mapSaplingNoteData.at(op);
-
-    auto optDeserialized = SaplingNotePlaintext::attempt_sapling_enc_decryption_deserialization(output.encCiphertext, nd.ivk, output.ephemeralKey);
-
-    // The transaction would not have entered the wallet unless
-    // its plaintext had been successfully decrypted previously.
-    assert(optDeserialized != std::nullopt);
-
-    auto maybe_pt = SaplingNotePlaintext::plaintext_checks_without_height(
-        *optDeserialized,
-        nd.ivk,
-        output.ephemeralKey,
-        output.cmu);
-    assert(maybe_pt != std::nullopt);
-    auto notePt = maybe_pt.value();
-
-    auto maybe_pa = nd.ivk.address(notePt.d);
-    assert(static_cast<bool>(maybe_pa));
-    auto pa = maybe_pa.value();
-
-    return std::make_pair(notePt, pa);
-}
-
-std::optional<std::pair<
-    SaplingNotePlaintext,
-    SaplingPaymentAddress>> CWalletTx::RecoverSaplingNote(const Consensus::Params& params, int height, SaplingOutPoint op, std::set<uint256>& ovks) const
-{
-    auto output = this->vShieldedOutput[op.n];
-
-    for (auto ovk : ovks) {
-        auto outPt = SaplingOutgoingPlaintext::decrypt(
-            output.outCiphertext,
-            ovk,
-            output.cv,
-            output.cmu,
-            output.ephemeralKey);
-        if (!outPt) {
-            // Try decrypting with the next ovk
-            continue;
-        }
-
-        auto maybe_pt = SaplingNotePlaintext::decrypt(
-            params,
-            height,
-            output.encCiphertext,
-            output.ephemeralKey,
-            outPt->esk,
-            outPt->pk_d,
-            output.cmu);
-        assert(static_cast<bool>(maybe_pt));
-        auto notePt = maybe_pt.value();
+    try {
+        auto decrypted = wallet::try_sapling_note_decryption(
+            *params.RustNetwork(),
+            // Canopy activation is inside the ZIP 212 grace period.
+            params.GetConsensus().vUpgrades[Consensus::UPGRADE_CANOPY].nActivationHeight,
+            nd.ivk.GetRawBytes(),
+            {
+                output.cv.GetRawBytes(),
+                output.cmu.GetRawBytes(),
+                output.ephemeralKey.GetRawBytes(),
+                output.encCiphertext,
+                output.outCiphertext,
+            });
 
-        return std::make_pair(notePt, SaplingPaymentAddress(notePt.d, outPt->pk_d));
+        return SaplingNotePlaintext::from_rust(std::move(decrypted));
+    } catch (const rust::Error &e) {
+        assert(false);
```

Comment:
Will the error `e` be logged and/or printed (besides causing the assertion which has no error info)? If not then please do so.